### PR TITLE
CellStatus connected to AppStatus

### PR DESCRIPTION
**Pull request by maackle**
This is a robust way to achieve the outcome of immediately Pausing an app whenever one of its cells fails to Join the network, whether during manual enabling or during manual or automatic restarting.

- Adds `reconcile_app_status_with_cell_status` and `reconcile_cell_status_with_app_status`, which coerces all app statuses to match the state of cells, and vice versa
- Uses cell status reconciliation to ensure that cells are properly added and removed whenever an app's status changes
    - Additionally, uses app status reconciliation when starting up an app, so that the app can be immediately paused if some cells failed to be created or fully join the network
- Uses cell status reconciliation to automatically remove "dangling" cells, i.e. cells which have no enabled app pointing to them.
    - This is used to remove cells when disabling an app, but also works in general

TODO:
- ensure kitsune can still interact with cells that are PendingJoin
- think if there are other places we need to do status reconciliation

### CODE REVIEW

File: crates/hc_sandbox/src/calls.rs [hunk 1]
```diff
@@ -563,7 +563,7 @@ pub async fn enable_app(cmd: &mut CmdRunner, args: EnableApp) -> anyhow::Result<
             installed_app_id: args.app_id,
         })
         .await?;
-    expect_match!(resp => AdminResponse::AppEnabled, format!("Failed to enable app, got: {:?}", resp));
+    matches::assert_matches!(resp, AdminResponse::AppEnabled { .. });
```
Comment (neonphog) [inline crates/hc_sandbox/src/calls.rs:566]: now that [matches](https://doc.rust-lang.org/std/macro.matches.html) is stable, we don't really need the external crate.

```suggestion
    assert!(matches!(resp, AdminResponse::AppEnabled { .. }));
```

File: crates/holochain/src/conductor/api/api_external/admin_interface.rs [hunk 2]
```diff
@@ -265,6 +280,7 @@ impl AdminInterfaceApi for RealAdminInterfaceApi {
                 Ok(AdminResponse::AppDisabled)
             }
             StartApp { installed_app_id } => {
+                // TODO: check to see if app was actually started
```
Comment (maackle) [inline crates/holochain/src/conductor/api/api_external/admin_interface.rs:283]: I think this is handled by the boolean in `AppStarted`, will double check

```suggestion
```

File: crates/hdk/Cargo.toml [hunk 3]
```diff
@@ -30,7 +30,7 @@ serde_bytes = "0.11"
 thiserror = "1.0.22"
 tracing = "0.1"
 tracing-core = "0.1"
-mockall = { version = "0.9", optional = true }
+mockall = { version = "0.8", optional = true }
```
Comment (neonphog) [inline crates/hdk/Cargo.toml:33]: why downgrading?

File: crates/holochain/src/conductor/api/api_cell.rs [hunk 4]
```diff
@@ -106,6 +106,7 @@ impl CellConductorApiT for CellConductorApi {
 }
 
 /// The "internal" Conductor API interface, for a Cell to talk to its calling Conductor.
+
```
Comment (neonphog) [inline crates/holochain/src/conductor/api/api_cell.rs:109]:
```suggestion
```

File: crates/holochain/src/conductor/conductor.rs [hunk 5]
```diff
@@ -61,15 +60,54 @@ use holochain_state::prelude::StateMutationResult;
...
 use tokio::sync::RwLock;
 use tracing::*;
 
 #[cfg(any(test, feature = "test_utils"))]
 use super::handle::MockConductorHandleT;
 
+/// Conductor-specific Cell state, this can probably be stored in a database.
+/// Hypothesis: If nothing remains in this struct, then the Conductor state is
+/// essentially immutable, and perhaps we just throw it out and make a new one
+/// when we need to load new config, etc.
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor.rs:74]:
```suggestion
/// The status of an installed Cell, which captures different phases of its lifecycle
```

File: crates/holochain/src/conductor/conductor.rs [hunk 6]
```diff
@@ -61,15 +60,54 @@ use holochain_state::prelude::StateMutationResult;
...
+    /// but it is considered not to be fully running from the perspective of
+    /// app status, i.e. if any app has a required Cell with this status,
+    /// the app is considered to be in the Paused state.
+    PendingJoin,
+}
+
+/// Declarative filter for CellStatus
+pub type CellStatusFilter = CellStatus;
+
 /// An [Cell] tracked by a Conductor, along with some [CellState]
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor.rs:90]:
```suggestion
/// A [`Cell`] tracked by a Conductor, along with its [`CellStatus`]
```

File: crates/holochain/src/conductor/conductor.rs [hunk 7]
```diff
@@ -151,11 +189,41 @@ where
...
+        let item = self
             .cells
             .get(cell_id)
             .ok_or_else(|| ConductorError::CellMissing(cell_id.clone()))?;
-        Ok(cell.0.clone())
+        Ok(item.cell.clone())
+    }
+
+    /// Iterator over all cells, regardless of status. Generally used when
+    /// responding to kitsune requests.
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor.rs:200]:
```suggestion
    /// Iterator over all cells, regardless of status.
```

File: crates/holochain/src/conductor/conductor.rs [hunk 8]
```diff
@@ -483,32 +551,87 @@ where
...
+                }
+            }
+            Ok(state)
+        })
+        .await?;
+        Ok(())
+    }
+
+    /// Remove all Cells which are not referenced by any Enabled app.
+    // TODO: test
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor.rs:607]:
```suggestion
    (Cells belonging to Paused apps are not considered "dangling" and will not be removed)
```

File: crates/holochain/src/conductor/cell.rs [hunk 9]
```diff
@@ -911,3 +911,9 @@ impl Cell {
         &self.queue_triggers
     }
 }
+
+impl std::fmt::Debug for Cell {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("Cell(id = {:?})", self.id()))
```
Comment (neonphog) [inline crates/holochain/src/conductor/cell.rs:917]: If you use the helper fns you get the `{:?}` vs `{:#?}` formatting for free:

```suggestion
        f.debug_struct("Cell").field("id", &self.id()).finish()
```

File: crates/holochain/src/conductor/conductor.rs [hunk 10]
```diff
@@ -517,98 +640,53 @@ where
...
-                                errors,
-                            })
-                        } else {
-                            // No errors so return the cells
-                            Ok(success.collect())
-                        }
-                    }
-                });
+
+        let app_cells: HashSet<&CellId> = state
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor.rs:644]:
```suggestion
        // Collect all CellIds across all apps, deduped
        let app_cells: HashSet<&CellId> = state
```

File: crates/holochain/src/conductor/conductor.rs [hunk 11]
```diff
@@ -664,11 +743,27 @@ where
...
+                CellItem {
+                    cell: Arc::new(cell),
+                    status: CellStatus::PendingJoin,
+                },
+            );
+            triggers.initialize_workflows();
+        }
+    }
+
+    /// Add fully constructed cells to the cell map in the Conductor
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor.rs:761]:
```suggestion
    /// Change the CellStatus of the given Cells in the Conductor. 
    /// Silently ignores Cells that don't exist.
```

File: crates/holochain/src/conductor/conductor.rs [hunk 12]
```diff
@@ -756,8 +851,11 @@ where
     /// Remove cells from the cell map in the Conductor
     pub(super) async fn remove_cells(&mut self, cell_ids: Vec<CellId>) {
         for cell_id in cell_ids {
-            if let Some(cell) = self.cells.remove(&cell_id) {
-                if let Err(err) = cell.0.cleanup().await {
+            if let Some(item) = self.cells.remove(&cell_id) {
+                // TODO: make this function async, or ensure that cleanup is fast,
+                //       so we don't hold the conductor lock unduly long while doing cleanup
+                //       (currently cleanup is a noop)
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor.rs:857]:
```suggestion
                //       (currently cleanup is a noop, so it's not urgent)
```

File: crates/holochain/src/conductor/conductor/tests.rs [hunk 13]
```diff
@@ -415,10 +419,19 @@ async fn test_uninstall_app() {
 async fn test_setup_cells_idempotency() {
```
Comment (maackle) [inline crates/holochain/src/conductor/conductor/tests.rs:419]:
```suggestion
async fn test_reconciliation_idempotency() {
```

File: crates/holochain_conductor_api/src/admin_interface.rs [hunk 14]
```diff
@@ -346,11 +346,14 @@ pub enum AdminResponse {
...
-    /// The succesful response to an [`AdminRequest::ListActiveApps`].
+    /// The succesful response to an [`AdminRequest::ListEnabledApps`].
     ///
     /// Contains a list of all the active `App` ids in the conductor
     ///
-    /// [`AdminRequest::ListActiveApps`]: enum.AdminRequest.html#variant.ListActiveApps
+    /// [`AdminRequest::ListEnabledApps`]: enum.AdminRequest.html#variant.ListEnabledApps
+    EnabledAppsListed(Vec<InstalledAppId>),
+
+    #[deprecated = "alias for EnabledAppsListed"]
```
Comment (neonphog) [inline crates/holochain_conductor_api/src/admin_interface.rs:356]: Does deprecating things work again? At one point we couldn't do this due to the `#[deny(warnings)]`...

**Review by neonphog — APPROVED** [review]
mostly cosmetic comments, looks good to me : )

File: crates/holochain/src/conductor/handle.rs [hunk 15]
```diff
@@ -673,30 +733,32 @@ impl<DS: DnaStore + 'static> ConductorHandleT for ConductorHandleImpl<DS> {
...
-            .flatten()
-            .collect();
-        Ok(r)
     }
 
+    #[tracing::instrument(skip(self))]
+    async fn reconcile_cell_status_with_app_status(
+        self: Arc<Self>,
+    ) -> ConductorResult<Vec<(CellId, CellError)>> {
+        // TODO: first leave the network for any cells that should be removed.
```
Comment (maackle) [inline crates/holochain/src/conductor/handle.rs:752]: NOTE, need to do this once merging in network leave

File: crates/hdk/Cargo.toml (hunk 3 above)
Comment (maackle) [reply crates/hdk/Cargo.toml:33]: Just to be uniform with the others. I tried upgrading to 0.10 but it broke everything so I just set them all back to 0.8

File: crates/holochain_conductor_api/src/admin_interface.rs (hunk 14 above)
Comment (maackle) [reply crates/holochain_conductor_api/src/admin_interface.rs:356]: If you don't use a deprecated thing anywhere, then there is no warning :) Although, we do have a bunch of `#![allow(deprecated)]` sprinkled around which we should eventually get around to removing...

**Comment by maackle** [comment]
Bad merge conflict, I have to go for today so I'll have to fix it tomorrow

**Pull request merged**